### ISSUE TITLE
Invisible signature fields and PDF/A-4 restrictions on flags, and what that means for PDF/UA-2

### ISSUE THREAD

**Issue by MatthiasValvekens** [pdf-association/pdf-issues#192]
**Describe the bug**

PDF/A-4 (like PDF/A-2 and PDF/A-3 before it) contains the following provision:

> Except for annotation dictionaries whose Subtype value is Popup, all annotation dictionaries shall contain the F key. If present, the F key’s Print flag bit shall be set to 1 and its Hidden, Invisible, ToggleNoView, and NoView flag bits shall be set to 0.

Note that invisible (=zero-size **Rect**) signature fields are _not_ exempt from this requirement! To understand why this is an issue, allow me to take a brief detour through PDF 1.7-land.

PDF/UA-1 contains the following provision in the general annotation requirements:

> The requirements of this clause shall not apply to annotations whose hidden flag is set or whose
rectangle is outside the CropBox or whose Subtype is Popup.

This means that, in order for an invisible signature field to be exempt from the tagging requirements in PDF/UA-1, it must have its hidden flag set (possible workaround: put them outside the cropbox?). Ergo, in a file that declares compliance with PDF/A-2 (or PDF/A-3) _and_ PDF/UA-1, all invisible signature fields are subject to tagging requirements.

In practice, this standard is impossible to meet: the classic example is PAdES-B-LTA document timestamp updates, which require a potentially unbounded number of signature fields to be added for timestamps. Under PDF/UA+PDF/A, that implies having to add all these document timestamp fields to the structure tree, which creates a royal mess for signature validation.

The effect is that it's impractical to use PAdES-B-LTA signatures effectively in PDF 1.7 documents that need to comply with both PDF/A and PDF/UA. I think we can all agree that that's not a good state of affairs.

--------------

Why does this story matter at all in a PDF 2.0 context? Because PDF/A-4 contains the same requirement on annotation flags as PDF/A-2 and PDF/A-3. Fortunately, widgets with zero width and height are considered artifacts in PDF/UA-2 (at least according to the current draft), so there's an escape hatch for (future) software seeking to comply with both. It's not particularly intuitive, though.

I would therefore (still) like to ask the PDF/A experts  here whether they would consider adding an exemption to 6.3.2 of ISO 19005-4 for "annotations where the value of the Rect key consists of an array where value 1 is equal to value 3 and value 2 is equal to value 4" (such an exemption does apply to appearance streams, interestingly enough).

Regardless of the outcome of that discussion, this clearly warrants a very loud, obnoxious note in PDF/UA-2... ;)

--------------

TL;DR: In current standards derived from PDF 1.7, the combined requirements of PDF/A and PDF/UA w.r.t. invisible signature fields are too onerous to enable modern (PAdES-style) digital signing workflows. Since PDF/A-4 did not change course on that issue, we either need to amend PDF/A-4 to allow the **Hidden** flag on invisible signature fields, or make it _crystal clear_ to future PDF/UA-2 users what they have to do. Ideally, both.

**Comment by MatthiasValvekens** [comment]
Pinging @bdoubrov and @DuffJohnson since this concerns alignment between PDF/A-4 and PDF/UA-2. :)

**Comment by bdoubrov** [comment]
I think the PDF/A requirement to have all four fields Hidden, Invisible, ToggleNoView, and NoView to be set to 0 is introduced to guarantee the consistency of different implementations in rendering (or not rendering) the appearance stream of the annotation. If there is no appearance stream (or if the annotation rectangle is empty) there is no risk here. So, yes, I don't see any harm extending the exception of "annotations where the value of the Rect key consists of an array where value 1 is equal to value 3 and value 2 is equal to value 4" to the Flags requirement as well.

**Comment by petervwyatt** [comment]
To be highly pedantic: Issue #124 corrected the array indices in ISO 32000-2:2020 to be zero-based, not 1-based in Table 166 (see https://pdf-issues.pdfa.org/32000-2-2020/clause12.html#Table166). So -1 from each of @bdoubrov indices above 😁

**Comment by lrosenthol** [comment]
Since PDF/A-4 is already out, and as you note, consistent with A-2 and A-3 - it would seem like a MUCH better solution to just fix this in UA-2 as you suggest.

**Comment by bdoubrov** [comment]
> To be highly pedantic: Issue #124 corrected the array indices in ISO 32000-2:2020 to be zero-based, not 1-based in Table 166 (see https://pdf-issues.pdfa.org/32000-2-2020/clause12.html#Table166). So -1 from each of @bdoubrov indices above 😁

Then we have to fix PDF/A-4 as well. As this is the wording used currently in the spec.

**Comment by MatthiasValvekens** [comment]
> I think the PDF/A requirement to have all four fields Hidden, Invisible, ToggleNoView, and NoView to be set to 0 is introduced to guarantee the consistency of different implementations in rendering (or not rendering) the appearance stream of the annotation. If there is no appearance stream (or if the annotation rectangle is empty) there is no risk here.

That's how I interpreted it as well. If we exempt these annotations from the appearance stream requirement, then we might as well allow them to set the Hidden flag. In a future PDF/A part, we could (?) even force that, just to be even more explicit.

> Since PDF/A-4 is already out, and as you note, consistent with A-2 and A-3 - it would seem like a MUCH better solution to just fix this in UA-2 as you suggest.

Yeah, I figured that that would be the default response... :) I concede that we can address the immediate problem by ensuring that zero-size annotations are artifacts in PDF/UA-2 (which is currently what the text says), but I think it would nevertheless behoove us to be consistent in how we define annotations as "invisible". The status quo is anything but:

 - PDF/A parts 2, 3, and 4 and ISO 32000-2 contain some exception provisions for the case where **Rect** has zero width/height, but still mandates the Print flag to be set (and Hidden unset) irrespective of that
 - PDF/UA-1 discriminates based on the value of the Hidden flag, but doesn't care about the size of **Rect** at all.

The problem I pointed out above is a direct consequence of that misalignment. I would propose that going forward, we should force these two definitions to agree in the subsets, or at least make sure they don't give rise to mutually incompatible provisions. If we can't "backport" those fixes to already-published specs, then so be it. Then we just have to make sure that PDF/UA-2 is sufficiently clear. But if we have the opportunity to address the discrepancy in (hypothetical) future parts of 19005, 14289 or 32000, we (IMHO) absolutely should straighten it out once and for all.

Either way, I'll try to draft a note to add to 8.10.3.4 in PDF/UA-2 to call out this use case, and submit it as a ballot comment against the current draft. For the record, I'm personally still in favour of relaxing the corresponding PDF/A-4 provision in a future revision of ISO 19005-4, but if others disagree, I'm content with ensuring that PDF/UA-2 gets it right.

**Comment by lrosenthol** [comment]
I understand that there may well be issues with PDF/A-2, 3 and 4....However, I given their target market (aka long term archives), I am not inclined to touch them in any way - even with an errata.  We've had too many issues in the past over disagreements over validation.

**Comment by lrosenthol** [comment]
That said...

Given the limited adoption of A-4, and the breath of errata on 32K-2, I *would* like to discuss the possibility of a dated revision at our fall meeting.

And same for PDF/X-6 at the 130 meeting

**Comment by petervwyatt** [comment]
Re PDF/A-4 errata: @bdoubrov and I have an action from the last WG5 meeting to table a list of errata from here at the next WG5 meeting so that WG5 can make a decision regarding amending or updating ISO 19005-4 - or not. Our deadline is mid-August-ish so getting something close to a proposed solution for this specific issue means it would make that list for consideration by WG5 SMEs...

**Comment by MatthiasValvekens** [comment]
Hi everyone, I've drafted some text.


## PDF/UA-2

Proposed note(s) for PDF/UA, in 8.9.2.3.13 (the clause on widget annotations, in case the numbering changes)

> NOTE 2: When used as part of a document workflow that makes use of signature profiles like PAdES-B-LTA, widget annotations associated with document timestamp signatures are often artifacts in this sense. In such workflows, the number of timestamps that will be added over time can typically not be predicted in advance. Since modifying the structure tree in incremental updates after signing complicates signature validation, it is important that invisible widgets be exempt from any tagging requirements otherwise imposed by this document.
> 
> NOTE 3: Invisible artifact widgets can nevertheless have their Print flag set, e.g. in cases where compliance with ISO 19005-4 (PDF/A-4) is required.

Obviously, NOTE 3 would be unnecessary if ISO 19005-4 were to be revised. But I think it's a good addition if we decide not to amend ISO 19005-4, since having to set Print on an otherwise invisible widget (that's also an artifact for UA purposes) is counterintuitive.

Wordsmith away!

--------------------------

## PDF/A-4

Here's the change I'd like to see in PDF/A-4, _if_ making that technical change is on the table at all.

In 6.3.2, change

> Except for annotation dictionaries whose Subtype value is Popup, all annotation dictionaries shall contain the F key. If present, the F key’s Print flag bit shall be set to 1 and its Hidden, Invisible, ToggleNoView, and NoView flag bits shall be set to 0.

to

> Except for annotation dictionaries whose Subtype value is Popup, and those where the value of the **Rect** key consists of an array where the value at index 0 is equal to the value at index 2 and the value at index 1 is equal to the value at index 3, all annotation dictionaries shall contain the F key. If present, the F key’s Print flag bit shall be set to 1 and its Hidden, Invisible, ToggleNoView, and NoView flag bits shall be set to 0.

I'd perhaps even go as far as to say that the Hidden flag _should_ be set on this type of zero-size annotation, but that's a separate discussion.

In my view, given the current state of adoption, these changes would be acceptable since they don't constrain any existing documents, and only require PDF/A-4 processors to accept more documents. Others may disagree, though, and I understand the reluctance towards making changes to a standard like PDF/A-4.


---------------------------

## Workaround in PDF 1.7

As a final comment on the status quo for PDF 1.7: after some experimentation, there's a somewhat hacky workaround that does clear the bar for all subsets. You can add the document timestamps without tagging them if you set the Print flag while making sure the bounding box (even if it has no area) doesn't intersect the crop box. Clearly this isn't particularly intuitive and also violates the intention behind the Print flag, but it's technically valid.

**Comment by MatthiasValvekens** [comment]
Great, thanks! On reflection, the first note might make more sense as an example, actually. Whichever you (+the editing committee) prefer :)

**Comment by MatthiasValvekens** [comment]
Update after WG9 meeting: the proposed notes were added to PDF/UA-2 (with some very minor modifications). This already solves the main problem, i.e. establishing a clear path to achieving A-4 + UA-2 compatibility in workflows that require multiple invisible signatures.

The proposed change for PDF/A-4 is still outstanding, and may or may not be adopted in a dated revision of PDF/A-4 in the future. I wouldn't close the issue just yet.

@petervwyatt @bdoubrov Maybe we can flag it for review in the PDF/A TWG? I'd also be OK with taking this up in the PDF TWG directly if you think there's nothing further to do here (other than punting it to WG5).

**Comment by petervwyatt** [comment]
Keeping this issue with the "PDF/A-4" label will ensure we will revisit this as the dated revision of PDF/A-4 moves forward.

**Comment by DuffJohnson** [comment]
In PDF/A TWG call the proposed change was NOT accepted, as it was pointed out that the language has existed for many years without raising any problems... from a PDF/A perspective. Accordingly, the PDF/A TWG would like the PDF/UA TWG to consider another approach for PDF/UA-2.

**Comment by lrosenthol** [comment]
@DuffJohnson @mrbhardy  can you check on PDF/UA-2?

**Comment by mrbhardy** [comment]
@lrosenthol I don't believe the problem exists in UA-2.  The previous text from UA-1 related to _Hidden_ no longer exists and we say any annotations can be an artifact.  We only specify that under certain circumstances, they shall be artifacts (but otherwise may or may not).

**Comment by petervwyatt** [comment]
PDF TWG note that this is not applicable to PDF/UA-2 and has already been addressed in PDF/A-4 dated revision.
@MatthiasValvekens - does this address all your concerns?

**Comment by MatthiasValvekens** [comment]
I don't have the text of the latest PDF/A-4 draft in front of me (only the pre-ballot version), but I saw in the RoV for the last CD that my suggestion was accepted in principle, so I trust the PL to have applied it.
AFAIK PDF/UA-2 also still includes the notes I suggested above to explain why it's important that invisible signatures can be considered artifacts.

So I think this one can be squared away--thanks!
